// @henrylabs-interview/payments/dist/index.js
import { Checkout } from './resources/checkout';
import { renderEmbeddedCheckout } from './resources/embedded';
import { Webhooks } from './resources/webhooks';
export class PaymentProcessor {
    checkout;
    webhooks;
    VALID_API_KEYS = ['824c951e-dfac-4342-8e03', '3f67e17a-880a-463b-a667', '78254d40-623a-48c5-b83f'];
    constructor(config) {
        if (!config.apiKey) {
            throw new Error('Henry: apiKey is required');
        }
        if (!this.VALID_API_KEYS.includes(config.apiKey)) {
            throw new Error('Henry: Invalid apiKey provided');
        }
        this.checkout = new Checkout();
        this.webhooks = new Webhooks();
    }
}
export class EmbeddedCheckout {
    checkoutId;
    constructor(config) {
        if (!config.checkoutId) {
            throw new Error('Henry: checkoutId is required');
        }
        this.checkoutId = config.checkoutId;
    }
    /**
     * Renders the embedded checkout UI
     * @param containerElementId - The ID of the container element where the checkout UI will be rendered
     * @param callbackFn - Callback function to handle the payment token
     * @returns - Whether the UI rendered successfully
     */
    render(containerElementId, callbackFn) {
        return renderEmbeddedCheckout(containerElementId, this.checkoutId, callbackFn);
    }
}
