// @henrylabs-interview/payments/dist/resources/embedded.js
import { isValidCardNumber, isValidExpiry } from '../utils/card';
import { generateTimeBasedID } from '../utils/crypto';
export async function renderEmbeddedCheckout(containerElementId, checkoutId, callbackFn) {
    // --- Environment Guard ---
    if (typeof window === 'undefined' || typeof document === 'undefined') {
        console.warn('EmbeddedCheckoutUI can only be used in a browser environment.');
        return false;
    }
    // --- Checkout Validation ---
    if (checkoutId.length !== 20 || !checkoutId.startsWith('cki_')) {
        console.warn(`Invalid checkout ID: "${checkoutId}".`);
        return false;
    }
    if (`${checkoutId}` !== `cki_${await generateTimeBasedID('checkout')}`) {
        console.warn(`Expired checkout ID: "${checkoutId}".`);
        return false;
    }
    const container = document.getElementById(containerElementId.startsWith('#') ? containerElementId.slice(1) : containerElementId);
    if (!container) {
        console.warn(`Container element "${containerElementId}" not found.`);
        return false;
    }
    // --- Render UI ---
    container.innerHTML = `
  <div style="
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
    background: #ffffff;
    border-radius: 16px;
    box-shadow: 0 10px 30px rgba(0,0,0,0.08);
    padding: 24px;
    max-width: 400px;
    width: 100%;
    box-sizing: border-box;
  ">
    <h3 style="margin:0 0 20px 0; font-size:18px; font-weight:600; color:#111827;">
      "Secure" Checkout
    </h3>

    <div style="display:flex; flex-direction:column; gap:14px;">

      ${inputBlock('card-number', 'Card Number', '4242 4242 4242 4242', 19)}
      <div style="display:flex; gap:10px;">
        <div style="flex:1;">
          ${inputBlock('exp-month', 'Exp. Month', '12', 2)}
        </div>
        <div style="flex:1;">
          ${inputBlock('exp-year', 'Exp. Year', '2030', 4)}
        </div>
      </div>
      ${inputBlock('cvc', 'CVC', '123', 3)}

      <button id="confirm-btn"
        style="
          margin-top:8px;
          padding:12px;
          border-radius:12px;
          border:none;
          font-size:14px;
          font-weight:600;
          background: linear-gradient(135deg, #6366f1, #4f46e5);
          color:white;
          cursor:pointer;
          transition: all 0.15s ease;
          box-shadow: 0 4px 14px rgba(79,70,229,0.3);
        ">
        Confirm Payment
      </button>

      <p style="font-size:11px; color:#9ca3af; text-align:center; margin-top:8px;">
        🔒 Payments are securely processed
      </p>

    </div>
  </div>
  `;
    function inputBlock(id, label, placeholder, maxLength) {
        return `
      <div style="display:flex; flex-direction:column;">
        <label style="font-size:12px; color:#6b7280; margin-bottom:6px;">
          ${label}
        </label>
        <input
          id="${id}"
          type="text"
		  inputmode="numeric"
		  maxlength="${maxLength}"
		  pattern="[0-9]*"
          placeholder="${placeholder}"
          style="
            padding:10px 12px;
            border-radius:10px;
            border:1px solid #e5e7eb;
            font-size:14px;
            outline:none;
            transition:border 0.2s, box-shadow 0.2s;
          "
        />
        <div id="${id}-error"
          style="font-size:12px; color:#ef4444; margin-top:4px; display:none;">
        </div>
      </div>
    `;
    }
    // --- DOM Refs ---
    const cardNumberInput = container.querySelector('#card-number');
    const expMonthInput = container.querySelector('#exp-month');
    const expYearInput = container.querySelector('#exp-year');
    const cvcInput = container.querySelector('#cvc');
    const button = container.querySelector('#confirm-btn');
    // --- Helpers ---
    function showError(input, message) {
        input.style.border = '1px solid #ef4444';
        const error = container?.querySelector(`#${input.id}-error`);
        if (error) {
            error.textContent = message;
            error.style.display = 'block';
        }
    }
    function clearError(input) {
        input.style.border = '1px solid #e5e7eb';
        const error = container?.querySelector(`#${input.id}-error`);
        if (error) {
            error.textContent = '';
            error.style.display = 'none';
        }
    }
    function setLoading(state) {
        button.disabled = state;
        button.style.opacity = state ? '0.6' : '1';
        button.textContent = state ? 'Processing...' : 'Confirm Payment';
    }
    function clearAllErrors() {
        [cardNumberInput, expMonthInput, expYearInput, cvcInput].forEach(clearError);
    }
    // --- Formatting ---
    cardNumberInput.addEventListener('input', () => {
        let value = cardNumberInput.value.replace(/\D/g, '').slice(0, 16);
        value = value.replace(/(.{4})/g, '$1 ').trim();
        cardNumberInput.value = value;
        clearError(cardNumberInput);
    });
    [expMonthInput, expYearInput, cvcInput].forEach((input) => input.addEventListener('input', () => clearError(input)));
    // --- Submit ---
    button.addEventListener('click', async () => {
        clearAllErrors();
        const number = cardNumberInput.value.replace(/\s+/g, '');
        const expMonth = Number(expMonthInput.value);
        const expYear = Number(expYearInput.value);
        const cvc = cvcInput.value;
        let hasError = false;
        if (!number) {
            showError(cardNumberInput, 'Card number required');
            hasError = true;
        }
        else if (!isValidCardNumber(number)) {
            showError(cardNumberInput, 'Invalid card number');
            hasError = true;
        }
        if (!expMonthInput.value || expMonth < 1 || expMonth > 12) {
            showError(expMonthInput, 'Invalid month');
            hasError = true;
        }
        if (!expYearInput.value || !isValidExpiry(expMonth, expYear)) {
            showError(expYearInput, 'Card expired');
            hasError = true;
        }
        if (!cvc || !/^\d{3,4}$/.test(cvc)) {
            showError(cvcInput, 'Invalid CVC');
            hasError = true;
        }
        if (hasError)
            return;
        setLoading(true);
        const paymentToken = `pmt_${await generateTimeBasedID('payment-token')}`;
        callbackFn(paymentToken);
    });
    return true;
}
